core/ir/riscv64/codec.c

```diff
@@ -1401,3 +1420,993 @@ decode_common(dcontext_t *dcontext, byte *pc, byte *orig_pc, instr_t *instr)
+ *        ^---^
+ * Applies to CL, CS, CA and CB compressed formats.
+ */
+static bool
+encode_crs1__opnd(instr_t *instr, byte *pc, int idx, uint32_t *out)
+{
+    opnd_t opnd = instr_get_src(instr, idx);
+    uint32_t rd = opnd_get_reg(opnd) - DR_REG_X8;
+    *out |= SET_FIELD(rd, 9, 7);
+    return true;
+}
+
+/* Encode the limited range (x8-x15) 2nd source fixed-point register field:
+ * |31  5|4    2|1      0|
+ * | ... | rs2' | opcode |
+ *        ^---^
+ * Applies to CS and CA compressed formats.
+ */
+static bool
+encode_crs2__opnd(instr_t *instr, byte *pc, int idx, uint32_t *out)
+{
+    opnd_t opnd = instr_get_src(instr, idx);
+    uint32_t rd = opnd_get_reg(opnd) - DR_REG_X8;
+    *out |= SET_FIELD(rd, 4, 2);
+    return true;
+}
+
+/* Encode the limited range (x8-x15) 2nd source floating-point register field:
+ * |31  5|4    2|1      0|
+ * | ... | rs2' | opcode |
+ *        ^---^
+ * Applies to CS and CA compressed formats.
+ */
+static bool
+encode_crs2_fp_opnd(instr_t *instr, byte *pc, int idx, uint32_t *out)
+{
+    opnd_t opnd = instr_get_src(instr, idx);
+    uint32_t rd = opnd_get_reg(opnd) - DR_REG_F8;
+    *out |= SET_FIELD(rd, 4, 2);
+    return true;
+}
+
+/* Encode the limited range (x8-x15) destination fixed-point register field:
+ * |31  5|4   2|1      0|
+ * | ... | rd' | opcode |
+ *        ^---^
+ * Applies to the CA compressed format.
+ */
+static bool
+encode_crd___opnd(instr_t *instr, byte *pc, int idx, uint32_t *out)
+{
+    opnd_t opnd = instr_get_dst(instr, idx);
+    uint32_t rd = opnd_get_reg(opnd) - DR_REG_X8;
+    *out |= SET_FIELD(rd, 9, 7);
+    return true;
+}
+
+/* Encode the 6-bit (6th bit always 0 in rv32) shift amount:
+ * |15    13|   12   |11    10|9    7|6        2|1      0|
+ * | funct3 | imm[5] | funct2 | rs1' | imm[4:0] | opcode |
+ *           ^------^                 ^--------^
+ */
+static bool
+encode_cshamt_opnd(instr_t *instr, byte *pc, int idx, uint32_t *out)
+{
+    opnd_t opnd = instr_get_src(instr, idx);
+    int32_t imm = opnd_get_immed_int(opnd);
+    *out |= SET_FIELD(imm, 6, 2) | SET_FIELD(imm >> 5, 12, 12);
+    return true;
+}
+
+/* Encode the CSR immediate in instructions from the Zicsr extension:
+ * |31 20|19      15|14  7|6      0|
+ * | csr | imm[4:0] | ... | opcode |
+ *        ^--------^
+ * From:
+ * |31  5|4        0|
+ * |  0  | imm[4:0] |
+ */
+static bool
+encode_csr_imm_opnd(instr_t *instr, byte *pc, int idx, uint32_t *out)
+{
+    opnd_t opnd = instr_get_src(instr, idx);
+    int32_t imm = opnd_get_immed_int(opnd);
+    *out |= SET_FIELD(imm, 19, 15);
+    return true;
+}
+
+/* Encode the immediate of the caddi16sp instruction:
+ * |15 13|   12   |11  7|6              2|1      0|
+ * | ... | imm[9] | ... | imm[4|6|8:7|5] | opcode |
+ *        ^------^       ^--------------^
+ * From:
+ * |31     9|8        4|3   0|
+ * | imm[9] | imm[8:4] |  0  |
+ */
+static bool
+encode_caddi16sp_imm_opnd(instr_t *instr, byte *pc, int idx, uint32_t *out)
```

Comment:
I think it will be worth spending the effort to add the tests in this PR itself. E.g. see suite/tests/api/ir_aarch64.c

Reply:
Agree, will do that.

Reply:
Seems a lot of work is needed to turn on tests (i.e. make `make test` work), so I think it's better to submit in another PR.

Reply:
> Seems a lot of work is needed to turn on tests (i.e. make make test work)

Do you mean you're seeing test failures or issues in getting the tests set up?

Ideally tests should go together with code as that helps in detecting issues that may be missed in review by  humans. We seem to be adding a lot of encode/decode changes in this PR, for which tests will be even more useful. I think adding a working unit test module will improve dev velocity in the long term.

Reply:
> Do you mean you're seeing test failures or issues in getting the tests set up?

Yes, when I run `make test`, most of the tests don't work. Some of them are caused by atomics that are fixed in https://github.com/DynamoRIO/dynamorio/pull/6089, some of them are not, but I haven't analyzed the cause. All tests are currently disabled in CI, so we don't see them failing.

I agree we should add unit tests for the encoder and decoder. But if I understand correctly, to do that, we need to get existing tests to work first. How about we keep this PR open, and I'll submit a new PR to make `make test` work first, after that merges, I can add the tests for this one. Sorry for the inconvenience.

Reply:
There is no need to get all existing tests to work before adding some unit tests here.  There are multiple approaches.  Are they currently building now and BUILD_TESTS is on and enable_testing() is being called, and the only issue is that many tests fail when executed?  Then use a CTest label to select just a subset of tests to run by default and only label the new unit test(s).  We are using labels to limit the tests for other not-fully-supported-yet platforms like Mac: see the `OSX` label in suite/tests/CMakeLists.txt, or the other labels there like `RUNS_ON_QEMU`.

Reply:
Thanks! Will try that.